ingress-watcher/pkg/watch/watch_test.go

```diff
@@ -0,0 +1,176 @@
+package watch
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"io/ioutil"
+	"net/http"
+	"testing"
+	"time"
+
+	"github.com/cybozu-go/well"
+	"github.com/cybozu/neco-containers/ingress-watcher/metrics"
+	"github.com/prometheus/client_golang/prometheus"
+	dto "github.com/prometheus/client_model/go"
+)
+
+const timeoutDuration = 550 * time.Millisecond
+
+const (
+	httpGetSuccessfulTotalName = "ingresswatcher_http_get_successful_total"
+	httpGetFailTotalName       = "ingresswatcher_http_get_fail_total"
+)
+
+func TestWatcherRun(t *testing.T) {
+	type fields struct {
+		targetURLs []string
+		interval   time.Duration
+		httpClient *http.Client
+	}
+	type args struct {
+		ctx context.Context
+	}
+
+	tests := []struct {
+		name   string
+		fields fields
+		result map[string]float64
+	}{
+		{
+			name: "GET success every 100ms in 550ms",
+			fields: fields{
+				targetURLs: []string{"foo", "bar"},
+				interval:   100 * time.Millisecond,
+				httpClient: newTestClient(func(req *http.Request) (*http.Response, error) {
+					return &http.Response{
+						StatusCode: http.StatusOK,
+						Body:       ioutil.NopCloser(bytes.NewBuffer([]byte(""))),
+						Header:     make(http.Header),
+					}, nil
+				}),
+			},
+			result: map[string]float64{
+				httpGetSuccessfulTotalName: 5,
+				httpGetFailTotalName:       0,
+			},
+		},
+
+		{
+			name: "GET fail every 100ms in 550ms",
+			fields: fields{
+				targetURLs: []string{"foo"},
+				interval:   100 * time.Millisecond,
+				httpClient: newTestClient(func(req *http.Request) (*http.Response, error) {
+					return nil, errors.New("error")
+				}),
+			},
+			result: map[string]float64{
+				httpGetSuccessfulTotalName: 0,
+				httpGetFailTotalName:       5,
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			registry := prometheus.NewRegistry()
+			metrics.HTTPGetSuccessfulTotal.Reset()
+			metrics.HTTPGetFailTotal.Reset()
+			registry.MustRegister(
+				metrics.HTTPGetSuccessfulTotal,
+				metrics.HTTPGetFailTotal,
+			)
+
+			// create watcher and run
+			w := NewWatcher(
+				tt.fields.targetURLs,
+				tt.fields.interval,
+				tt.fields.httpClient,
+			)
+			env := well.NewEnvironment(context.Background())
+			env.Go(func(ctx context.Context) error {
+				ctx, cancel := context.WithTimeout(ctx, timeoutDuration)
+				defer cancel()
+				return w.Run(ctx)
+			})
+			env.Stop()
+			env.Wait()
```

Comment:
Why do not you check error retuned by `env.Wait()`?

Reply:
If the error is checked here, the error caused by timeout is also captured (the timeout behavior is intended).
I think the timeout error can be separated from the other errors, but I don't do that because of simplicity.